### PR TITLE
Feature/preview and extraction

### LINKED ISSUE #11: Calculate column-level stats
- [ ] Provide stats calculation status(pending, complete)
- [ ] Use GQL Subscription feature
- [ ] Column 
## stats
- min
- max
- count
- sum
- null count
- range
- unique values count
- NA counts
- standard deviation, mean, median
- Default aggregation

Image

Image

### LINKED ISSUE #13: Import table and Background Extraction Task
- [x] Import API
- [x] Extract table from sheet based on table properties

### LINKED ISSUE #20: Table preview API
- [x] Extract preview data
- [x] Table Properties mutation
- [x] Show updated preview after properties mutation
- [x] Add row key to each preview row item.

### PR DESCRIPTION
**Pull request by bewakes**
Resolves #13 
Resolves #20 
Resolves #11 

Depends on: None

## Changes

* Background task for table extraction
* Calculation of column stats

Mention related users here if any.

## This PR doesn't introduce any:

- [x] temporary files, auto-generated files or secret keys
- [x] n+1 queries
- [x] flake8 issues
- [x] `print`
- [x] typos
- [x] unwanted comments

## This PR contains valid:

- [x] tests
- [ ] permission checks (tests here too)
- [ ] translations

### CODE REVIEW

File: utils/extraction.py [hunk 1]
```diff
@@ -1,15 +1,18 @@
-from typing import Tuple, Optional, Union, cast, Any, TypedDict, List, Type
+from typing import Tuple, Optional, Union, cast, Any, TypedDict, List, Type, Dict
```
Comment (k9845) [inline utils/extraction.py:1]:
```suggestion
from typing import (
    Tuple,
    Optional,
    Union,
    cast,
     Any,
     TypedDict,
     List,
     Type,
     Dict,
)
```

File: utils/extraction.py [hunk 2]
```diff
@@ -43,23 +78,105 @@ def extract_preview_data(
...
+def calculate_stats_for_numeric_col(items: list):
+    # TODO: optimize the list(use np/pd). But this should happen from the extraction phase itself
+    return {
+        "min": float(np.min(items)),
+        "max": float(np.max(items)),
+        "mean": float(np.mean(items)),
+        "median": float(np.median(items)),
+        "std_deviation": float(np.std(items)),
+        "total_count": len(items),
+        "na_count": len([x for x in items if x is None]),
```
Comment (k9845) [inline utils/extraction.py:153]:
```suggestion
        "na_count": len([i for i in items if np.isnan(i) == True]),
```

File: utils/extraction.py [hunk 3]
```diff
@@ -43,23 +78,105 @@ def extract_preview_data(
...
+        if length > max_len:
+            max_len = length
+        if length < min_len:
+            min_len = length
+
+    # TODO: optimize the following calculations
+    return {
+        "total_count": len(items),
+        "na_count": len([x for x in items if x is None]),
+        "unique_count": len(set(items)),
```
Comment (k9845) [inline utils/extraction.py:173]: @bewakes how about using numpy for all starts calculations?

**Review by k9845 — CHANGES_REQUESTED** [review]

File: utils/extraction.py (hunk 3 above)
Comment (bewakes) [reply utils/extraction.py:173]: That's the ultimate plan. The calculations themselves need to be optimized and are naively done at the moment. We'll work on these after the first MVP.

File: utils/extraction.py (hunk 2 above)
Comment (bewakes) [reply utils/extraction.py:153]: The thing is, during the extraction, na values are converted to `None` as we need to save those as json in database and nan/numpy values are not JSON serializable. Thus, `None` is the way to go here.

File: utils/extraction.py (hunk 1 above)
Comment (bewakes) [reply utils/extraction.py:1]: Hmm. good point. I expected `black` to do these kind of formatting. Seems like it does not do so. will fix.

File: apps/core/models.py [hunk 4]
```diff
@@ -86,7 +100,6 @@ class ActionType(models.TextChoices):
         TABLE_WISE = "table-wise", _("Table Wise")
 
     table = models.ForeignKey(Table, null=True, on_delete=models.SET_NULL)
-    column = models.ForeignKey(Column, null=True, on_delete=models.SET_NULL)
     snapshot_version = models.CharField(max_length=20)
```
[the remark this answers is not in the document]
Comment (bewakes) [reply apps/core/models.py:103]: Hmm. that's a very good point. I am thinking maybe we won't even need a reference to snapshot here. This model is not used and final yet. We'll have to think about this more after we complete the APIs for table rows and column stats. Let's leave it as it is for now.

File: apps/core/models.py [hunk 5]
```diff
@@ -32,14 +32,15 @@ class TableStatus(models.TextChoices):
         PENDING = "pending", _("Pending")
         EXTRACTED = "extracted", _("Extracted")
 
+    original_name = models.CharField(max_length=255)
```
[the remark this answers is not in the document]
Comment (bewakes) [reply apps/core/models.py:36]: Good point. Done/

File: apps/core/models.py [hunk 6]
```diff
@@ -58,26 +59,39 @@ def clone(self):
...
+    def data_rows(self):
+        if self.last_snapshot is None:
+            return []
+        return self.last_snapshot.data_rows
+
+    @property
+    def data_columns(self):
+        if self.last_snapshot is None:
+            return []
+        return self.last_snapshot.data_rows
```
[the remark this answers is not in the document]
Comment (bewakes) [reply apps/core/models.py]: Thanks for pointing that out!

**Review by k9845 — APPROVED** [review]

**Pull request merged**